ota_proxy/ota_cache.py

```diff
@@ -35,9 +36,25 @@ def _subprocess_check_output(cmd: str, *, raise_exception=False) -> str:
         return ""
 
 
+class _RegisterDict(dict):
+    def register(self, url: str) -> bool:
+        if url in self:
+            return False
+        else:
+            self[url] = None
+            return True
+
+    def unregister(self, url: str):
+        try:
+            del self[url]
+        except KeyError:
+            pass
```

Comment:
I have question, this member is accessed from multiple threads? (if so lock is required)

Reply:
with GIL, operating set itself is thread safe. However, since `check url exists` and `add url in the set` in `register` method are not a single atomic operation, it may be edge condition happens, but it will not impact the file caching, so decided to not use lock.

Reply:
Entering here from multiple threads OK?
https://github.com/tier4/ota-client/pull/113/files#diff-bf7539f67c1e8add5ca7cafafc3886db722d3e80105caa19c4d9ce99584e4e1cR615-R620

```
                    # failed to get the chance to cache
                    fp, meta = await self._open_fp_by_requests(
                        url, cookies, extra_headers
                    )
                    res = OTAFile(url=url, fp=fp, meta=meta)
```

Reply:
> Entering here from multiple threads OK? https://github.com/tier4/ota-client/pull/113/files#diff-bf7539f67c1e8add5ca7cafafc3886db722d3e80105caa19c4d9ce99584e4e1cR615-R620
> 
> ```
>                     # failed to get the chance to cache
>                     fp, meta = await self._open_fp_by_requests(
>                         url, cookies, extra_headers
>                     )
>                     res = OTAFile(url=url, fp=fp, meta=meta)
> ```

Yes, it will not cause problems.
Actually the edge condition is when 2 threads simultaneously successfully get the chance to cache file, that will not cause problem, but just result in a file being cached twice. New copy will override old copy.

Reply:
If this cache operation `just result in a file being cached twice.`, is safe operation, I think it is OK,
I think at least a comment is necessary, but I think it is easier to understand by adding a lock than by reading that comment.
Why don't you add lock?

Reply:
When I read above comment that `lock is not required since...`, I think I'll try to submit to add lock and remove that comment.

Reply:
> If this cache operation `just result in a file being cached twice.`, is safe operation, I think it is OK, I think at least a comment is necessary, but I think it is easier to understand by adding a lock than by reading that comment. Why don't you add lock?

I try the implementation with lock, the performance is suffered obviously. Since we don't get too much improvement from adding lock, I decided to not add lock. In short, the reason is performance penalty.

Reply:
OK, Could you add those comments?

Reply:
Sure!